README.md
# `tar-no-std-extended` - Parse Tar Archives (Tarballs)

_Due to historical reasons, there are several formats of tar archives. All of them are based on the same principles,
but have some subtle differences that often make them incompatible with each other._ [[0]]

Library to read Tar archives (by GNU Tar) in `no_std` contexts with zero allocations. If you have a standard
environment and need full feature support, I recommend the use of <https://crates.io/crates/tar> instead.

## Limitations
The crate is simple and only supports reading of "basic" archives, therefore no extensions, such
as GNU Longname. The maximum supported file name length is 256 characters excluding the NULL-byte (using the tar name/prefix longname implementation). The maximum supported file size is 8GiB. Directories are supported, but only regular fields are yielded in iteration.


## Use Case

This library is useful, if you write a kernel or a similar low-level application, which needs
"a bunch of files" from an archive ("init ramdisk"). The Tar file could for example come
as a Multiboot2 boot module provided by the bootloader.

This crate focuses on extracting files from uncompressed Tar archives created with default options by **GNU Tar**.
GNU Extensions such as sparse files, incremental archives, and long filename extension are not supported yet.
[This link](https://www.gnu.org/software/tar/manual/html_section/Formats.html) gives a good overview over possible
archive formats and their limitations.

## Example (without `alloc`-feature)
```rust
use tar_no_std_extended::TarArchiveRef;

fn main() {
    // log: not mandatory
    std::env::set_var("RUST_LOG", "trace");
    env_logger::init();

    // also works in no_std environment (except the println!, of course)
    let archive = include_bytes!("../tests/gnu_tar_default.tar");
    let archive = TarArchiveRef::new(archive);
    // Vec needs an allocator of course, but the library itself doesn't need one
    let entries = archive.entries().collect::<Vec<_>>();
    println!("{:#?}", entries);
    println!("content of last file:");
    println!("{:#?}", entries[2].data_as_str().expect("Should be valid UTF-8"));
}
```

## Alloc Feature
This crate allows the usage of the additional Cargo build time feature `alloc`. When this is used,
the crate also provides the type `TarArchive`, which owns the data on the heap.

## Compression (`tar.gz`)
If your tar file is compressed, e.g. by `.tar.gz`/`gzip`, you need to uncompress the bytes first
(e.g. by a *gzip* library). Afterwards, this crate can read the Tar archive format from the uncompressed
bytes.

## MSRV
The MSRV is 1.52.1 stable.


[0]: https://www.gnu.org/software/tar/manual/html_section/Formats.html


CHANGELOG.md
# v0.3.0 (2023-08-29) - Andy Schneider
- Handle directories.
- Fix defect in header type declaration.
- Fix defect handling octal numbers with historical space padding.
- Add support for long name (name + prefix) support.
- Add header iterator.
- Internal test improvements.

# v0.2.0 (2023-04-11)
- MSRV is 1.60.0
- bitflags bump: 1.x -> 2.x
- few internal code improvements (less possible panics)
- `Mode::to_flags` now returns a Result
- Feature `all` was removed. Use `alloc` instead.


src/header.rs
/*
MIT License

Copyright (c) 2023 Philipp Schuster
Copyright (c) 2023 Andy Schneider

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
//! TAR header definition taken from <https://www.gnu.org/software/tar/manual/html_node/Standard.html>.
//! A Tar-archive is a collection of 512-byte sized blocks. Unfortunately there are several
//! TAR-like archive specifications. An Overview can be found here:
//! <https://www.gnu.org/software/tar/manual/html_node/Formats.html#Formats>
//!
//! This library focuses on extracting files from the GNU Tar format.

#![allow(non_upper_case_globals)]

use crate::{BLOCKSIZE, NAME_LEN, PREFIX_LEN, TarFormatString, TarFormatOctal, TarFormatDecimal};
use core::fmt::{Debug, Formatter};
use core::num::ParseIntError;

#[derive(Debug)]
pub enum ModeError {
    ParseInt(ParseIntError),
    IllegalMode,
}

/// Wrapper around the UNIX file permissions given in octal ASCII.
#[derive(Copy, Clone)]
#[repr(transparent)]
pub struct Mode(TarFormatOctal<8>);

impl Mode {
    /// Parses the [`ModeFlags`] from the mode string.
    pub fn to_flags(self) -> Result<ModeFlags, ModeError> {
        let bits =
            self.0.as_number::<u64>().map_err(ModeError::ParseInt)?;
        ModeFlags::from_bits(bits).ok_or(ModeError::IllegalMode)
    }
}

impl Debug for Mode {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        let mut debug = f.debug_tuple("Mode");
        debug.field(&self.to_flags());
        debug.finish()
    }
}

/// Header of the TAR format as specified by POSIX (POSIX 1003.1-1990.
/// "New" (version?) GNU Tar versions use this archive format by default.
/// (<https://www.gnu.org/software/tar/manual/html_node/Formats.html#Formats>).
///
/// Each file is started by such a header, that describes the size and
/// the file name. After that, the file content stands in chunks of 512 bytes.
/// The number of bytes can be derived from the file size.
///
/// This is also mostly compatible with the "Ustar"-header and the "GNU format".
/// Because this library only needs to fetch data and filename, we don't need
/// further checks.
#[derive(Debug, Copy, Clone)]
#[repr(C, packed)]
pub struct PosixHeader {
    pub name: TarFormatString<NAME_LEN>,
    pub mode: Mode,
    pub uid: TarFormatOctal<8>,
    pub gid: TarFormatOctal<8>,
    // confusing; size is stored as ASCII string
    pub size: TarFormatOctal<12>,
    pub mtime: TarFormatDecimal<12>,
    pub cksum: TarFormatOctal<8>,
    pub typeflag: TypeFlag,
    /// Name. There is always a null byte, therefore
    /// the max len is 99.
    pub linkname: TarFormatString<NAME_LEN>,
    pub magic: TarFormatString<6>,
    pub version: TarFormatString<2>,
    /// Username. There is always a null byte, therefore
    /// the max len is N-1.
    pub uname: TarFormatString<32>,
    /// Groupname. There is always a null byte, therefore
    /// the max len is N-1.
    pub gname: TarFormatString<32>,
    pub dev_major: TarFormatOctal<8>,
    pub dev_minor: TarFormatOctal<8>,
    pub prefix: TarFormatString<PREFIX_LEN>,
    // padding => to BLOCKSIZE bytes
    pub _pad: [u8; 12],
}

impl PosixHeader {
    /// Returns the number of blocks that are required to read the whole file
    /// content. Returns an error, if the file size can't be parsed from the
    /// header.
    pub fn payload_block_count(&self) -> Result<usize, ParseIntError> {
        let div = self.size.as_number::<usize>()? / BLOCKSIZE;
        let modulo = self.size.as_number::<usize>()? % BLOCKSIZE;
        let block_count = if modulo > 0 { div + 1 } else { div };
        Ok(block_count)
    }

    /// A Tar archive is terminated, if a end-of-archive entry, which consists of two 512 blocks
    /// of zero bytes, is found.
    pub fn is_zero_block(&self) -> bool {
        let ptr = self as *const Self as *const u8;
        let self_bytes = unsafe { core::slice::from_raw_parts(ptr, BLOCKSIZE) };
        self_bytes.iter().filter(|x| **x == 0).count() == BLOCKSIZE
    }
}

/// Describes the kind of payload, that follows after a
/// [`PosixHeader`]. The properties of this payload are
/// described inside the header.
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
#[repr(u8)]
#[allow(unused)]
pub enum TypeFlag {
    /// Represents a regular file. In order to be compatible with older versions of tar, a typeflag
    /// value of AREGTYPE should be silently recognized as a regular file. New archives should be
    /// created using REGTYPE. Also, for backward compatibility, tar treats a regular file whose
    /// name ends with a slash as a directory.
    REGTYPE = b'0',
    /// Represents a regular file. In order to be compatible with older versions of tar, a typeflag
    /// value of AREGTYPE should be silently recognized as a regular file. New archives should be
    /// created using REGTYPE. Also, for backward compatibility, tar treats a regular file whose
    /// name ends with a slash as a directory.
    AREGTYPE = b'\0',
    /// This flag represents a file linked to another file, of any type, previously archived. Such
    /// files are identified in Unix by each file having the same device and inode number. The
    /// linked-to name is specified in the linkname field with a trailing null.
    LINK = b'1',
    /// This represents a symbolic link to another file. The linked-to name is specified in the
    /// linkname field with a trailing null.
    SYMTYPE = b'2',
    /// Represents character special files and block special files respectively. In this case the
    /// devmajor and devminor fields will contain the major and minor device numbers respectively.
    /// Operating systems may map the device specifications to their own local specification, or
    /// may ignore the entry.
    CHRTYPE = b'3',
    /// Represents character special files and block special files respectively. In this case the
    /// devmajor and devminor fields will contain the major and minor device numbers respectively.
    /// Operating systems may map the device specifications to their own local specification, or
    /// may ignore the entry.
    BLKTYPE = b'4',
    /// This flag specifies a directory or sub-directory. The directory name in the name field
    /// should end with a slash. On systems where disk allocation is performed on a directory
    /// basis, the size field will contain the maximum number of bytes (which may be rounded to
    /// the nearest disk block allocation unit) which the directory may hold. A size field of zero
    /// indicates no such limiting. Systems which do not support limiting in this manner should
    /// ignore the size field.
    DIRTYPE = b'5',
    /// This specifies a FIFO special file. Note that the archiving of a FIFO file archives the
    /// existence of this file and not its contents.
    FIFOTYPE = b'6',
    /// This specifies a contiguous file, which is the same as a normal file except that, in
    /// operating systems which support it, all its space is allocated contiguously on the disk.
    /// Operating systems which do not allow contiguous allocation should silently treat this type
    /// as a normal file.
    CONTTYPE = b'7',
    /// Extended header referring to the next file in the archive
    XHDTYPE = b'x',
    /// Global extended header
    XGLTYPE = b'g',
}

bitflags::bitflags! {
    /// UNIX file permissions in octal format.
    #[repr(transparent)]
    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    pub struct ModeFlags: u64 {
        /// Set UID on execution.
        const SetUID = 0o4000;
        /// Set GID on execution.
        const SetGID = 0o2000;
        /// Reserved.
        const TSVTX = 0o1000;
        /// Owner read.
        const OwnerRead = 0o400;
        /// Owner write.
        const OwnerWrite = 0o200;
        /// Owner execute.
        const OwnerExec = 0o100;
        /// Group read.
        const GroupRead = 0o040;
        /// Group write.
        const GroupWrite = 0o020;
        /// Group execute.
        const GroupExec = 0o010;
        /// Others read.
        const OthersRead = 0o004;
        /// Others read.
        const OthersWrite = 0o002;
        /// Others execute.
        const OthersExec = 0o001;
    }
}

#[cfg(test)]
mod tests {
    use crate::header::{PosixHeader, TypeFlag};
    use crate::BLOCKSIZE;
    use std::mem::size_of;

    /// Casts the bytes to a reference to a PosixhHeader.
    fn bytes_to_archive(bytes: &[u8]) -> &PosixHeader {
        unsafe { (bytes.as_ptr() as *const PosixHeader).as_ref() }.unwrap()
    }

    #[test]
    fn test_display_header() {
        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_default.tar"));
        println!("{:#?}'", archive);
    }

    #[test]
    fn test_show_tar_header_magics() {
        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_default.tar"));
        println!(
            "default: magic='{:?}', version='{:?}'",
            archive.magic, archive.version
        );
        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_gnu.tar"));
        println!(
            "gnu: magic='{:?}', version='{:?}'",
            archive.magic, archive.version
        );
        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_oldgnu.tar"));
        println!(
            "oldgnu: magic='{:?}', version='{:?}'",
            archive.magic, archive.version
        );
        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_pax.tar"));
        println!(
            "pax: magic='{:?}', version='{:?}'",
            archive.magic, archive.version
        );
        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_posix.tar"));
        println!(
            "posix: magic='{:?}', version='{:?}'",
            archive.magic, archive.version
        );
        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_ustar.tar"));
        println!(
            "ustar: magic='{:?}', version='{:?}'",
            archive.magic, archive.version
        );
        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_v7.tar"));
        println!(
            "v7: magic='{:?}', version='{:?}'",
            archive.magic, archive.version
        );
    }

    #[test]
    fn test_parse_tar_header_filename() {
        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_default.tar"));
        assert_eq!(
            archive.typeflag,
            TypeFlag::REGTYPE,
            "the first entry is a regular file!"
        );
        assert_eq!(archive.name.as_str(), "bye_world_513b.txt");

        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_gnu.tar"));
        assert_eq!(
            archive.typeflag,
            TypeFlag::REGTYPE,
            "the first entry is a regular file!"
        );
        assert_eq!(archive.name.as_str(), "bye_world_513b.txt");

        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_oldgnu.tar"));
        assert_eq!(
            archive.typeflag,
            TypeFlag::REGTYPE,
            "the first entry is a regular file!"
        );
        assert_eq!(archive.name.as_str(), "bye_world_513b.txt");

        /* UNSUPPORTED YET. Uses extensions..
        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_pax.tar"));
        assert_eq!(archive.typeflag, TypeFlag::REGTYPE, "the first entry is a regular file!");
        assert_eq!(archive.name.as_string().as_str(), "bye_world_513b.txt"); */

        /* UNSUPPORTED YET. Uses extensions.
        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_posix.tar"));
        unsupported extension XHDTYPE assert_eq!(archive.typeflag, TypeFlag::REGTYPE, "the first entry is a regular file!");
        assert_eq!(archive.name.as_string().as_str(), "bye_world_513b.txt"); */

        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_ustar.tar"));
        assert_eq!(
            archive.typeflag,
            TypeFlag::REGTYPE,
            "the first entry is a regular file!"
        );
        assert_eq!(archive.name.as_str(), "bye_world_513b.txt");

        let archive = bytes_to_archive(include_bytes!("../tests/gnu_tar_v7.tar"));
        // ARegType: legacy
        assert_eq!(
            archive.typeflag,
            TypeFlag::AREGTYPE,
            "the first entry is a regular file!"
        );
        assert_eq!(archive.name.as_str(), "bye_world_513b.txt");
    }

    #[test]
    fn test_size() {
        assert_eq!(BLOCKSIZE, size_of::<PosixHeader>());
    }
}


examples/minimal.rs
/*
MIT License

Copyright (c) 2023 Philipp Schuster
Copyright (c) 2023 Andy Schneider

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
use tar_no_std_extended::TarArchiveRef;

fn main() {
    // log: not mandatory
    std::env::set_var("RUST_LOG", "trace");
    env_logger::init();

    // also works in no_std environment (except the println!, of course)
    let archive = include_bytes!("../tests/gnu_tar_default.tar");
    let archive = TarArchiveRef::new(archive);
    // Vec needs an allocator of course, but the library itself doesn't need one
    let entries = archive.entries().collect::<Vec<_>>();
    println!("{:#?}", entries);
    println!("content of last file:");
    println!(
        "{:#?}",
        entries[2].data_as_str().expect("Should be valid UTF-8")
    );
}


src/tar_format_types.rs
#![allow(unused_imports)]

/// An optionally null terminated string. The contents are either:
/// 1. A fully populated string with no null termination or
/// 2. A partially populated string where the unused bytes are zero.
/// 
/// The content is likely to be UTF-8/ASCII, but that is not verified by this
/// type.
#[derive(Copy, Clone)]
#[repr(C)]
pub struct TarFormatString<const N: usize> {
    bytes: [u8; N]
}

/// A number. Trailing spaces in the string are ignored.
#[derive(Copy, Clone)]
#[repr(C)]
pub struct TarFormatNumber<const N: usize, const R: u32>(TarFormatString<N>);

/// An octal number. Trailing spaces in the string are ignored.
#[derive(Copy, Clone)]
#[repr(C)]
pub struct TarFormatOctal<const N: usize>(TarFormatNumber<N, 8>);

/// A decimal number. Trailing spaces in the string are ignored.
#[derive(Copy, Clone)]
#[repr(C)]
pub struct TarFormatDecimal<const N: usize>(TarFormatNumber<N, 10>);

use core::fmt::{Debug, Formatter};
use core::num::ParseIntError;
use core::str::from_utf8;
use core::ptr::copy_nonoverlapping;
use num_traits::Num;

impl<const N: usize> TarFormatString<N> {
    /// Constructor.
    pub fn new(bytes: [u8; N]) -> Self {
        if N == 0 {
            panic!("Array cannot be zero length");
        }
        Self { bytes }
    }

    /// True if the string empty (ignoring NULL bytes).
    pub const fn is_empty(&self) -> bool {
        self.bytes[0] == 0
    }

    // True if the string is NULL terminated
    pub const fn is_nul_terminated(&self) -> bool {
        return self.bytes[N-1] == 0;
    }

    /// Returns the length of the string (ignoring NULL bytes).
    pub fn len(&self) -> usize {
        if self.is_nul_terminated() {
            memchr::memchr(0, &self.bytes).unwrap()
        } else {
            N
        }
    }

    /// Returns a str ref without NULL bytes
    pub fn as_str(&self) -> &str
    {
        from_utf8(&self.bytes[0..self.len()]).expect("byte array is not UTF-8")
    }

    /// Append to end of string
    pub fn append<const S: usize>(&mut self, other: &TarFormatString<S>) {
        let resulting_length = self.len() + other.len();
        if resulting_length > N {
            panic!("Result to long for capacity {}", N);
        }

        unsafe {
            let dst = self.bytes.as_mut_ptr().add(self.len());
            let src = other.bytes.as_ptr();
            copy_nonoverlapping(src, dst, other.len());
            if resulting_length < N {
                self.bytes[resulting_length] = 0;
            }
        }
    }
}

impl<const N: usize> Debug for TarFormatString<N> {
    fn fmt(&self, f: &mut Formatter) -> core::fmt::Result {
        let sub_array = &self.bytes[0 .. self.len()];
        write!(f, "{},{} of {},{}", from_utf8(sub_array).unwrap(), self.len(), N, self.is_nul_terminated())
    }
}

impl<const N: usize, const R: u32> TarFormatNumber<N, R> {
    pub fn as_number<T>(&self) -> core::result::Result<T, T::FromStrRadixErr> where 
    T: num_traits::Num {
        match memchr::memchr2(32, 0, &self.0.bytes) {
            None => T::from_str_radix(self.0.as_str(), R),
            Some(idx) => {
                let sub_str = from_utf8(&self.0.bytes[..idx]).expect("byte array is not UTF-8");
                T::from_str_radix(sub_str, 8)
            }
        }
    }
}

impl<const N: usize, const R: u32> Debug for TarFormatNumber<N, R> {
    fn fmt(&self, f: &mut Formatter) -> core::fmt::Result {
        let sub_array = &self.0.bytes[0 .. self.0.len()];
        match self.as_number::<u64>() {
            Err(msg) => write!(f, "{} [{}]", msg, from_utf8(sub_array).unwrap()),
            Ok(val) => write!(f, "{} [{}]", val, from_utf8(sub_array).unwrap())
        }
    }
}


impl<const N: usize> Debug for TarFormatOctal<N> {
    fn fmt(&self, f: &mut Formatter) -> core::fmt::Result {
        self.0.fmt(f)
    }
}

impl<const N: usize> Debug for TarFormatDecimal<N> {
    fn fmt(&self, f: &mut Formatter) -> core::fmt::Result {
        self.0.fmt(f)
    }
}

impl<const N: usize> TarFormatDecimal<N> {
    pub fn as_number<T>(&self) -> core::result::Result<T, T::FromStrRadixErr> where 
    T: num_traits::Num {
        self.0.as_number::<T>()
    }
}

impl<const N: usize> TarFormatOctal<N> {
    pub fn as_number<T>(&self) -> core::result::Result<T, T::FromStrRadixErr> where 
    T: num_traits::Num {
        self.0.as_number::<T>()
    }
}


mod tests
{
    use super::TarFormatString;

    use core::mem::size_of_val;

    #[test]
    fn test_empty_string()
    {
        let empty= TarFormatString::new([0]);
        assert_eq!(size_of_val(&empty), 1);
        assert!(empty.is_empty());
        assert_eq!(empty.len(), 0);
        assert!(empty.is_nul_terminated());
        assert_eq!(empty.as_str(), "");
    }

    #[test]
    fn test_one_byte_string()
    {
        let s = TarFormatString::new([65]);
        assert_eq!(size_of_val(&s), 1);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 1);
        assert!(!s.is_nul_terminated());
        assert_eq!(s.as_str(), "A");
    }

    #[test]
    fn test_two_byte_string_nul_terminated()
    {
        let s = TarFormatString::new([65, 0]);
        assert_eq!(size_of_val(&s), 2);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 1);
        assert!(s.is_nul_terminated());
        assert_eq!(s.as_str(), "A");
    }

    #[test]
    fn test_append()
    {
        let mut s = TarFormatString::new([0;20]);

        // When adding a zero terminated string with one byte of zero
        s.append(&TarFormatString::new([0]));
        // Then the result is no change
        assert_eq!(size_of_val(&s), 20);
        assert!(s.is_empty());
        assert_eq!(s.len(), 0);
        assert!(s.is_nul_terminated());
        assert_eq!(s.as_str(), "");

        s.append(&TarFormatString::new([65, 66, 67]));
        assert_eq!(size_of_val(&s), 20);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 3);
        assert!(s.is_nul_terminated());
        assert_eq!(s.as_str(), "ABC");

        s.append(&TarFormatString::new([68, 69, 70]));
        assert_eq!(size_of_val(&s), 20);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 6);
        assert!(s.is_nul_terminated());
        assert_eq!(s.as_str(), "ABCDEF");

        assert_eq!(s.len() + 14, 20);

        s.append(&TarFormatString::new([b'A'; 12]));
        assert_eq!(size_of_val(&s), 20);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 18);
        assert!(s.is_nul_terminated());
        assert_eq!(s.as_str(), "ABCDEFAAAAAAAAAAAA");

        s.append(&TarFormatString::new([b'A'; 1]));
        assert_eq!(size_of_val(&s), 20);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 19);
        assert!(s.is_nul_terminated());
        assert_eq!(s.as_str(), "ABCDEFAAAAAAAAAAAAA");

        s.append(&TarFormatString::new([b'Z'; 1]));
        assert_eq!(size_of_val(&s), 20);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 20);
        assert!(!s.is_nul_terminated());
        assert_eq!(s.as_str(), "ABCDEFAAAAAAAAAAAAAZ");
    }    

}

src/tar_format_string.rs
#![allow(unused_imports)]

/// An optionally null terminated string. The contents are either:
/// 1. A fully populated string with no null termination or
/// 2. A partially populated string where the unused bytes are zero.


/// A C-String that is stored in a static array. There is always a terminating
/// NULL-byte.
///
/// The content is likely to be UTF-8/ASCII, but that is not verified by this
/// type.
#[derive(Copy, Clone)]
#[repr(C)]
pub struct TarFormatString<const N: usize> {
    bytes: [u8; N]
}

use core::fmt::{Debug, Formatter, Result};
use core::str::from_utf8;
use core::ptr::copy_nonoverlapping;

impl<const N: usize> TarFormatString<N> {
    /// Constructor.
    pub fn new(bytes: [u8; N]) -> Self {
        if N == 0 {
            panic!("Array cannot be zero length");
        }
        Self {bytes }
    }

    /// Returns if the string empty (ignoring NULL bytes).
    pub const fn is_empty(&self) -> bool {
        self.bytes[0] == 0
    }

    // Returns if the string is NULL terminated
    pub const fn is_nul_terminated(&self) -> bool {
        return self.bytes[N-1] == 0;
    }

    /// Returns the length of the string (ignoring NULL bytes).
    pub fn len(&self) -> usize {
        if self.is_nul_terminated() {
            memchr::memchr(0, &self.bytes).unwrap()
        } else {
            N
        }
    }

    /// Returns a str ref without NULL bytes
    pub fn as_str(&self) -> &str
    {
        from_utf8(&self.bytes[0..self.len()]).expect("byte array is not UTF-8")
    }

    /// Append to end of string
    pub fn append<const S: usize>(&mut self, other: &TarFormatString<S>) {
        let resulting_length = self.len() + other.len();
        if resulting_length > N {
            panic!("Result to long for capacity {}", N);
        }

        unsafe {
            let dst = self.bytes.as_mut_ptr().add(self.len());
            let src = other.bytes.as_ptr();
            copy_nonoverlapping(src, dst, other.len());
            if resulting_length < N {
                self.bytes[resulting_length] = 0;
            }
        }
    }
}

impl<const N: usize> Debug for TarFormatString<N> {
    fn fmt(&self, f: &mut Formatter) -> Result {
        let sub_array = &self.bytes[0 .. self.len()];
        write!(f, "{},{} of {},{}", from_utf8(sub_array).unwrap(), self.len(), N, self.is_nul_terminated())
    }
}

mod tests
{
    use super::TarFormatString;

    use core::mem::size_of_val;

    #[test]
    fn test_empty_string()
    {
        let empty= TarFormatString::new([0]);
        assert_eq!(size_of_val(&empty), 1);
        assert!(empty.is_empty());
        assert_eq!(empty.len(), 0);
        assert!(empty.is_nul_terminated());
        assert_eq!(empty.as_str(), "");
    }

    #[test]
    fn test_one_byte_string()
    {
        let s = TarFormatString::new([65]);
        assert_eq!(size_of_val(&s), 1);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 1);
        assert!(!s.is_nul_terminated());
        assert_eq!(s.as_str(), "A");
    }

    #[test]
    fn test_two_byte_string_nul_terminated()
    {
        let s = TarFormatString::new([65, 0]);
        assert_eq!(size_of_val(&s), 2);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 1);
        assert!(s.is_nul_terminated());
        assert_eq!(s.as_str(), "A");
    }

    #[test]
    fn test_append()
    {
        let mut s = TarFormatString::new([0;20]);

        // When adding a zero terminated string with one byte of zero
        s.append(&TarFormatString::new([0]));
        // Then the result is no change
        assert_eq!(size_of_val(&s), 20);
        assert!(s.is_empty());
        assert_eq!(s.len(), 0);
        assert!(s.is_nul_terminated());
        assert_eq!(s.as_str(), "");

        s.append(&TarFormatString::new([65, 66, 67]));
        assert_eq!(size_of_val(&s), 20);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 3);
        assert!(s.is_nul_terminated());
        assert_eq!(s.as_str(), "ABC");

        s.append(&TarFormatString::new([68, 69, 70]));
        assert_eq!(size_of_val(&s), 20);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 6);
        assert!(s.is_nul_terminated());
        assert_eq!(s.as_str(), "ABCDEF");

        assert_eq!(s.len() + 14, 20);

        s.append(&TarFormatString::new([b'A'; 12]));
        assert_eq!(size_of_val(&s), 20);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 18);
        assert!(s.is_nul_terminated());
        assert_eq!(s.as_str(), "ABCDEFAAAAAAAAAAAA");

        s.append(&TarFormatString::new([b'A'; 1]));
        assert_eq!(size_of_val(&s), 20);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 19);
        assert!(s.is_nul_terminated());
        assert_eq!(s.as_str(), "ABCDEFAAAAAAAAAAAAA");

        s.append(&TarFormatString::new([b'Z'; 1]));
        assert_eq!(size_of_val(&s), 20);
        assert!(!s.is_empty());
        assert_eq!(s.len(), 20);
        assert!(!s.is_nul_terminated());
        assert_eq!(s.as_str(), "ABCDEFAAAAAAAAAAAAAZ");
    }    

}

src/lib.rs
/*
MIT License

Copyright (c) 2023 Philipp Schuster
Copyright (c) 2023 Andy Schneider

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
//! Library to read Tar archives (by GNU Tar) in `no_std` contexts with zero
//! allocations. If you have a standard environment and need full feature
//! support, I recommend the use of <https://crates.io/crates/tar> instead.
//!
//! The crate is simple and only supports reading of "basic" archives, therefore
//! no extensions, such as GNU Longname. The maximum supported file name length
//! is 100 characters including the NULL-byte. The maximum supported file size
//! is 8 GiB. Also, directories are not supported yet but only flat collections
//! of files.
//!
//! This library is useful, if you write a kernel or a similar low-level
//! application, which needs "a bunch of files" from an archive ("init ram
//! disk"). The Tar file could for example come as a Multiboot2 boot module
//! provided by the bootloader.
//!
//! This crate focuses on extracting files from uncompressed Tar archives
//! created with default options by **GNU Tar**. GNU Extensions such as sparse
//! files, incremental archives, and long filename extension are not supported
//! yet. [gnu.org](https://www.gnu.org/software/tar/manual/html_section/Formats.html)
//! provides a good overview over possible archive formats and their
//! limitations.
//!
//! # Example
//! ```rust
//! use tar_no_std_extended::TarArchiveRef;
//!
//! // log: not mandatory
//! std::env::set_var("RUST_LOG", "trace");
//! env_logger::init();
//!
//! // also works in no_std environment (except the println!, of course)
//! let archive = include_bytes!("../tests/gnu_tar_default.tar");
//! let archive = TarArchiveRef::new(archive);
//! // Vec needs an allocator of course, but the library itself doesn't need one
//! let entries = archive.entries().collect::<Vec<_>>();
//! println!("{:#?}", entries);
//! println!("content of last file:");
//! let last_file_content = unsafe { core::str::from_utf8_unchecked(entries[2].data()) };
//! println!("{:#?}", last_file_content);
//! ```

#![cfg_attr(not(test), no_std)]
#![deny(
    clippy::all,
    clippy::cargo,
    clippy::nursery,
    // clippy::restriction,
    // clippy::pedantic
)]
// now allow a few rules which are denied by the above statement
// --> they are ridiculous and not necessary
#![allow(
    clippy::suboptimal_flops,
    clippy::redundant_pub_crate,
    clippy::fallible_impl_from
)]
#![deny(missing_debug_implementations)]
#![deny(rustdoc::all)]

#[cfg_attr(test, macro_use)]
#[cfg(test)]
extern crate std;

#[cfg(feature = "alloc")]
extern crate alloc;

/// Each Archive Entry (either Header or Data Block) is a block of 512 bytes.
const BLOCKSIZE: usize = 512;
/// Maximum filename length of the base Tar format including the terminating NULL-byte.
const NAME_LEN: usize = 100;
/// Maximum long filename length of the base Tar format including the prefix
const POSIX_1003_MAX_FILENAME_LEN: usize = 256;
/// Maximum length of the prefix in Posix tar format
const PREFIX_LEN: usize = 155;

mod archive;
mod header;
mod tar_format_types;

pub use archive::*;
pub use header::*;
pub use tar_format_types::*;
